File: frontend/CodeInTasks/src/app/web-api/services/task.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { TaskServiceInterface } from '../interfaces';
import { TaskCreateModel, TaskCreateResultModel, TaskFilterModel, TaskUpdateModel, TaskViewModel } from '../models';

@Injectable()
export class TaskService implements TaskServiceInterface {
  private static readonly basePath = "/api/task";

  constructor(private httpClient: HttpClient) { }

  public getAsync(taskId: string): Observable<TaskViewModel> {
    return this.httpClient.get<TaskViewModel>(`${TaskService.basePath}/${taskId}`);
  }

  public getFilteredAsync(filterModel: TaskFilterModel): Observable<TaskViewModel[]> {
    return this.httpClient.get<TaskViewModel[]>(TaskService.basePath, { params: <any>filterModel });
  }

  public addAsync(createModel: TaskCreateModel): Observable<TaskCreateResultModel> {
    return this.httpClient.post<TaskCreateResultModel>(TaskService.basePath, createModel);
  }

  public updateAsync(taskId: string, updateModel: TaskUpdateModel): Observable<unknown> {
    return this.httpClient.put(`${TaskService.basePath}/${taskId}`, updateModel);
  }

  public deleteAsync(taskId: string): Observable<unknown> {
    return this.httpClient.delete(`${TaskService.basePath}/${taskId}`);
  }
}

File: backend/WebApi/src/CodeInTasks.Application/Services/TaskService.cs
using AutoMapper;
using CodeInTasks.WebApi.Models.Task;

namespace CodeInTasks.Application.Services
{
    internal class TaskService : ITaskService
    {
        private readonly IRepository<TaskModel> taskRepository;
        private readonly IFiltrationPipeline<TaskFilterModel, TaskModel> filtrationPipeline;
        private readonly IMapper mapper;

        public TaskService(
            IRepository<TaskModel> taskRepository,
            IFiltrationPipeline<TaskFilterModel, TaskModel> filtrationPipeline,
            IMapper mapper)
        {
            this.taskRepository = taskRepository;
            this.filtrationPipeline = filtrationPipeline;
            this.mapper = mapper;
        }

        public Task<TaskCreateResultModel> AddAsync(TaskCreateModel taskCreateModel)
        {
            var taskModel = mapper.Map<TaskModel>(taskCreateModel);

            var resultTask = taskRepository.AddAsync(taskModel)
                .ContinueWith(addTask =>
                {
                    var taskId = addTask.Result;

                    return new TaskCreateResultModel()
                    {
                        TaskId = taskId,
                    };
                });

            return resultTask;
        }

        public async Task DeleteAsync(Guid taskId)
        {
            var isDeleted = await taskRepository.DeleteAsync(taskId);

            if (!isDeleted)
            {
                throw new EntityNotFoundException(nameof(TaskModel), taskId);
            }
        }

        public Task<IEnumerable<TaskModel>> GetFilteredAsync(TaskFilterModel filterModel)
        {
            var pipelineResult = filtrationPipeline.GetResult(filterModel);
            var filter = new RepositoryFilter<TaskModel>()
            {
                FiltrationPredicate = pipelineResult.FilterExpression,
                OrderFunction = pipelineResult.OrderFunction,
                Take = filterModel.TakeCount,
                Skip = filterModel.TakeOffset
            };

            var resultTask = taskRepository.GetFilteredAsync(filter);

            return resultTask;
        }

        public Task<TaskModel> GetAsync(Guid taskId)
        {
            var resultTask = GetTaskAsync(taskId);

            return resultTask;
        }

        public async Task UpdateAsync(TaskUpdateModel taskUpdateModel)
        {
            var taskId = taskUpdateModel.TaskId;
            var taskModel = await GetTaskAsync(taskId);

            mapper.Map(taskUpdateModel, taskModel);

            await taskRepository.UpdateAsync(taskModel);
        }

        public Task<bool> IsOwnerAsync(Guid taskId, Guid userId)
        {
            var resultTask = taskRepository.AnyAsync(x => x.Id == taskId && x.CreatorId == userId);

            return resultTask;
        }

        private async Task<TaskModel> GetTaskAsync(Guid taskId)
        {
            var taskModel = await taskRepository.GetAsync(taskId);

            return taskModel ?? throw new EntityNotFoundException(nameof(TaskModel), taskId);
        }
    }
}
